pkg/client/k8s/v1alpha1/unstructured.go

```diff
@@ -218,45 +216,137 @@ func UpdateLabels(o UnstructuredOptions) UnstructuredMiddleware {
+	// NotPredicates represents NOT logical operator
+	NotPredicates PredicateListOp = "not"
+)
+
+// UnstructPredicateList represents a list of predicates with a logical operator
+// to evaluate the condition
+type UnstructPredicateList struct {
+	Op    PredicateListOp
+	Items map[string]UnstructuredPredicate
+}
+
+func (l UnstructPredicateList) String() string {
+	var names []string
+	for name := range l.Items {
+		names = append(names, name)
+	}
+	return fmt.Sprintf("unstruct predicate list: op='%s', items='%s'", l.Op, strings.Join(names, ","))
+}
+
+// isNot does a NOT logic among the predicate list against the given
+// unstructured instance
+func (l UnstructPredicateList) isNot(u *unstructured.Unstructured) bool {
+	if u == nil {
+		return false
+	}
+	for _, predicate := range l.Items {
+		if predicate(u) {
+			return false
+		}
+	}
+	return true
 }
 
-// MapAll will execute the all UnstructuredMiddlewares on each unstructured
+// isAll does a AND logic among the predicate list against the given
+// unstructured instance
+func (l UnstructPredicateList) isAll(u *unstructured.Unstructured) bool {
+	if u == nil {
+		return false
+	}
+	for _, predicate := range l.Items {
+		if !predicate(u) {
+			return false
+		}
+	}
+	return true
+}
+
+// isAny does a OR logic among the predicate list against the given unstructured
 // instance
-func (u UnstructList) MapAll(ml []UnstructuredMiddleware) (ul UnstructList) {
-	for _, unstruct := range u.Items {
-		for _, m := range ml {
-			unstruct = m(unstruct)
+func (l UnstructPredicateList) isAny(u *unstructured.Unstructured) bool {
+	if u == nil {
+		return false
+	}
+	for _, predicate := range l.Items {
+		if predicate(u) {
+			return true
+		}
+	}
+	return false
+}
+
+// UnstructList represents a list of unstructured instances
+type UnstructList []*unstructured.Unstructured
+
+// MapAll executes all unstructured middlewares on each unstructured
+// instance and returns a new list of these updated unstructured instances
+func (u UnstructList) MapAll(ml UnstructuredMiddlewareList) (l UnstructList) {
+	if len(u) == 0 {
+		return
+	}
+	for _, unstruct := range u {
+		if unstruct == nil {
+			continue
 		}
-		if unstruct != nil {
-			ul.Items = append(ul.Items, unstruct)
+		updated := ml.Update(unstruct)
+		if updated != nil {
+			l = append(l, updated)
 		}
 	}
 	return
 }
 
-// MapIf will execute the UnstructuredMiddleware conditionally based on
-// UnstructuredPredicate
-func (u UnstructList) MapIf(m UnstructuredMiddleware, p UnstructuredPredicate) (ul UnstructList) {
-	for _, unstruct := range u.Items {
-		if p(unstruct) {
-			unstruct = m(unstruct)
+// Filter filters the unstructured instances based on the given predicates
+func (u UnstructList) Filter(pl UnstructPredicateList) (l UnstructList) {
+	if len(u) == 0 {
+		return
+	}
+	for _, unstruct := range u {
+		if unstruct == nil {
+			continue
 		}
-		if unstruct != nil {
-			ul.Items = append(ul.Items, unstruct)
+		switch pl.Op {
+		case AllPredicates:
```

Comment:
can we reorder this and maybe remove continue statement to make it more readable.